sql/plan/external_procedure.go

```diff
@@ -0,0 +1,187 @@
+	return true
+}
+
+// RowIter implements the interface sql.Node.
+func (n *ExternalProcedure) RowIter(ctx *sql.Context, row sql.Row) (sql.RowIter, error) {
+	// The function's structure has been verified by the analyzer, so no need to double-check any of it here
+	funcVal := reflect.ValueOf(n.Function)
+	funcType := funcVal.Type()
+	// The first parameter is always the context, but it doesn't exist as far as the stored procedures are concerned, so
+	// we prepend it here
+	funcParams := make([]reflect.Value, len(n.Params)+1)
+	funcParams[0] = reflect.ValueOf(ctx)
+
+	for i, exprParam := range n.Params {
+		funcParamType := funcType.In(i + 1)
+		// Grab the passed-in variable and convert it to the type we expect
+		exprParamVal, err := exprParam.Eval(ctx, nil)
+		if err != nil {
+			return nil, err
+		}
+		exprParamVal, err = n.ParamDefinitions[i].Type.Convert(exprParamVal)
+		if err != nil {
+			return nil, err
+		}
+
+		funcParamCompType := funcParamType
+		if funcParamType.Kind() == reflect.Ptr {
+			funcParamCompType = funcParamType.Elem()
+		}
+		// Convert to bool, []byte, int, and uint as they differ from their sql.Type value
+		switch funcParamCompType {
+		case boolType:
+			val := false
+			if exprParamVal.(int8) != 0 {
+				val = true
+			}
+			exprParamVal = val
+		case byteSliceType:
+			exprParamVal = []byte(exprParamVal.(string))
+		case intType:
+			if strconv.IntSize == 32 {
+				exprParamVal = int(exprParamVal.(int32))
+			} else {
+				exprParamVal = int(exprParamVal.(int64))
+			}
+		case uintType:
+			if strconv.IntSize == 64 {
+				exprParamVal = int(exprParamVal.(uint32))
+			} else {
+				exprParamVal = int(exprParamVal.(uint64))
+			}
+		case decimalType:
+			exprParamVal, err = decimal.NewFromString(exprParamVal.(string))
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		if funcParamType.Kind() == reflect.Ptr { // Coincides with INOUT
+			funcParamVal := reflect.New(funcParamType.Elem())
+			funcParamVal.Elem().Set(reflect.ValueOf(exprParamVal))
+			funcParams[i+1] = funcParamVal
+		} else { // Coincides with IN
+			funcParamVal := reflect.New(funcParamType)
+			funcParamVal.Elem().Set(reflect.ValueOf(exprParamVal))
+			funcParams[i+1] = funcParamVal.Elem()
+		}
+	}
+	out := funcVal.Call(funcParams)
+
+	// Again, these types are enforced in the analyzer, so it's safe to assume their types here
+	if err, ok := out[1].Interface().(error); ok { // Only evaluates to true when error is not nil
+		return nil, err
+	}
+	for i, paramDefinition := range n.ParamDefinitions {
+		if paramDefinition.Direction == ProcedureParamDirection_Inout || paramDefinition.Direction == ProcedureParamDirection_Out {
+			exprParam := n.Params[i]
+			funcParamVal := funcParams[i+1].Elem().Interface()
+			err := exprParam.Set(funcParamVal, exprParam.Type())
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	if rowIter, ok := out[0].Interface().(sql.RowIter); ok {
+		return rowIter, nil
+	}
+	return sql.RowsToRowIter(), nil
```

Comment:
Should be an error, right?

Reply:
I wouldn't say so. If you've no rows to return, seems valid to return nothing rather than an empty iter.